util/topsql/reporter/single_target.go

```diff
@@ -20,43 +20,97 @@ import (
 	"sync"
 	"time"
 
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/util"
 	"github.com/pingcap/tidb/util/logutil"
 	"github.com/pingcap/tipb/go-tipb"
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/backoff"
 )
 
-// ReportClient send data to the target server.
-type ReportClient interface {
-	Send(ctx context.Context, addr string, data reportData) error
-	Close()
-}
-
-// GRPCReportClient reports data to grpc servers.
-type GRPCReportClient struct {
+// SingleTargetDataSink reports data to grpc servers.
+type SingleTargetDataSink struct {
 	curRPCAddr string
 	conn       *grpc.ClientConn
 	// calling decodePlan this can take a while, so should not block critical paths
 	decodePlan planBinaryDecodeFunc
+	sendTaskCh chan sendTask
 }
 
-// NewGRPCReportClient returns a new GRPCReportClient
-func NewGRPCReportClient(decodePlan planBinaryDecodeFunc) *GRPCReportClient {
-	return &GRPCReportClient{
+// NewSingleTargetDataSink returns a new SingleTargetDataSink
+func NewSingleTargetDataSink(decodePlan planBinaryDecodeFunc) *SingleTargetDataSink {
+	dataSink := &SingleTargetDataSink{
 		decodePlan: decodePlan,
+		sendTaskCh: make(chan sendTask, 1),
 	}
+	go util.WithRecovery(dataSink.run, nil)
+	return dataSink
 }
 
-var _ ReportClient = &GRPCReportClient{}
+// run will return when ds.sendTaskCh is closed.
+func (ds *SingleTargetDataSink) run() {
+	for task := range ds.sendTaskCh {
+		targetRPCAddr := config.GetGlobalConfig().TopSQL.ReceiverAddress
+		if targetRPCAddr == "" {
+			continue
+		}
 
-// Send implements the ReportClient interface.
-// Currently the implementation will establish a new connection every time, which is suitable for a per-minute sending period
-func (r *GRPCReportClient) Send(ctx context.Context, targetRPCAddr string, data reportData) error {
-	if targetRPCAddr == "" {
-		return nil
+		ctx, cancel := context.WithDeadline(context.Background(), task.deadline)
+		start := time.Now()
+		err := ds.doSend(ctx, targetRPCAddr, task.data)
+		cancel()
+		if err != nil {
+			logutil.BgLogger().Warn("[top-sql] single target data sink failed to send data to receiver", zap.Error(err))
+			reportAllDurationFailedHistogram.Observe(time.Since(start).Seconds())
+		} else {
+			reportAllDurationSuccHistogram.Observe(time.Since(start).Seconds())
+		}
+	}
+}
+
+var _ DataSink = &SingleTargetDataSink{}
+
+// Send implements the DataSink interface.
+// Currently the implementation will establish a new connection every time,
+// which is suitable for a per-minute sending period
+func (ds *SingleTargetDataSink) Send(data ReportData, deadline time.Time) {
```

Comment:
Will it be better to call this interface `TrySend` to indicate that it will discard things? (I'm not sure! Please check your PubSub data sink as well to see whether this behavior is correct)

Reply:
PubSub will act the same, so `TrySend` is reasonable.